SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosOfferScheduler.java

```diff
@@ -579,7 +578,12 @@ private double score(SingularityOfferHolder offerHolder, SingularityTaskRequestH
     final SlaveMatchState slaveMatchState = slaveAndRackManager.doesOfferMatch(offerHolder, taskRequest, activeTaskIdsForRequest, isPreemptibleTask(taskRequest, deployStatsCache));
```

Comment:
I think your overall strategy here is good, but we need to be a bit more efficient. We already iterate over allowed/required attributes during this doesOfferMatch call here. Maybe we can communicate back a flag in the `SlaveMatchState` that lets this code know to boost the score? That way we don't have to do the attribute comparison and calculation a second time. Instead of just `SlaveMatchState.OK` could be `SlaveMatchState.PREFERRED_SLAVE` or something like that